classification/models/vessel_characterization.py

```diff
@@ -0,0 +1,231 @@
+import json
+from .model import ModelBase
+from . import layers
+from classification import metadata
+from .objectives import (
+    TrainNetInfo, MultiClassificationObjective, LogRegressionObjectiveMAE)
+from classification.feature_generation import vessel_feature_generation
+import logging
+import math
+import numpy as np
+import os
+
+import tensorflow as tf
+
+approx_means = [ 6.9139094e+00,  4.3506036e+00,  6.1346573e-01,  5.7922113e-01,
+                -1.0719098e-03, -3.7928432e-02,  6.0396111e-03, -2.2346964e-01,
+                -1.9947174e-01, -8.9961719e-03,  1.5873306e+00,  0.0000000e+00,
+                 0.0000000e+00,  0.0000000e+00]
+approx_stds =  [0.8707472 , 3.2674177 , 0.7677098 , 0.80428886, 0.3563173 ,
+                0.7140226 , 0.6984951 , 0.68056667, 0.6656215 , 0.54217416,
+                1.8521472 , 0.        , 0.        , 0.        ]
+
+class Model(ModelBase):
+
+    window_size = 3
+    feature_depths = [48, 64, 96, 128, 192, 256, 384, 512, 768]
+    strides = [2] * 9
+    assert len(strides) == len(feature_depths)
+    feature_sub_depths = 1024
+
+    initial_learning_rate = 100e-5
+    learning_decay_rate = 0.5
+    decay_examples = 100000
+
+    @property
+    def number_of_steps(self):
+        return 800000
+
+    @property
+    def max_window_duration_seconds(self):
+        return 180 * 24 * 3600
+
+    @property
+    def window_max_points(self):
+        nominal_max_points = (self.max_window_duration_seconds / (5 * 60)) / 4
+        layer_reductions = np.prod(self.strides)
+        final_size = int(round(nominal_max_points / layer_reductions))
+        max_points = final_size * layer_reductions
+        logging.info('Using %s points', max_points)
+        return max_points
+
+    @property
+    def min_viable_timeslice_length(self):
+        return 500
+
+    def __init__(self, num_feature_dimensions, vessel_metadata, metrics):
+        super(Model, self).__init__(num_feature_dimensions, vessel_metadata)
+
+        class XOrNan:
+            def __init__(self, key):
+                self.key = key
+
+            def __call__(self, id_):
+                x = vessel_metadata.vessel_label(self.key, id_)
+                if x == '':
+                    x = np.nan
+                return np.float32(x)
+
+        self.training_objectives = [
+            LogRegressionObjectiveMAE(
+                'length',
+                'Vessel-length',
+                XOrNan('length'),
+                metrics=metrics,
+                loss_weight=0.1),
+            LogRegressionObjectiveMAE(
+                'tonnage',
+                'Vessel-tonnage',
+                XOrNan('tonnage'),
+                metrics=metrics,
+                loss_weight=0.1),
+            LogRegressionObjectiveMAE(
+                'engine_power',
+                'Vessel-engine-Power',
+                XOrNan('engine_power'),
+                metrics=metrics,
+                loss_weight=0.1),
+            LogRegressionObjectiveMAE(
```

Comment:
Interesting. How does this perform?

Reply:
Not obviously better or worse, unfortunately.